utils/oscap-info.c

```diff
@@ -174,17 +261,254 @@ static inline void _print_xccdf_benchmark(struct xccdf_benchmark *bench, const c
+{
+	struct xccdf_profile_iterator *prof_it = xccdf_benchmark_get_profiles(bench);
+	_print_xccdf_profile_with_id(prof_it, profile_id, "", &_print_xccdf_profile_verbose);
+	xccdf_profile_iterator_free(prof_it);
+}
+
+static void _print_single_benchmark_all(struct xccdf_benchmark *bench, const char *prefix)
+{
+	_print_xccdf_benchmark(bench, prefix, 0);
+}
+
+static int app_info_single_ds_profiles_only(struct ds_stream_index_iterator* sds_it, struct ds_sds_session *session, const struct oscap_action *action)
+{
+	struct ds_stream_index * stream = ds_stream_index_iterator_next(sds_it);
+	struct oscap_string_iterator* checklist_it = ds_stream_index_get_checklists(stream);
+
+	while (oscap_string_iterator_has_more(checklist_it)) {
+		const char * id = oscap_string_iterator_next(checklist_it);
+		/* decompose */
+		struct oscap_source *xccdf_source = ds_sds_session_select_checklist(session, ds_stream_index_get_id(stream), id, NULL);
+		if (xccdf_source == NULL) {
+			oscap_string_iterator_free(checklist_it);
+			ds_stream_index_iterator_free(sds_it);
+			ds_sds_session_free(session);
+			return OSCAP_ERROR;
+		}
+
+		if (oscap_source_get_scap_type(xccdf_source) == OSCAP_DOCUMENT_XCCDF) {
+			struct xccdf_benchmark* bench = xccdf_benchmark_import_source(xccdf_source);
+			if(!bench) {
+				oscap_string_iterator_free(checklist_it);
+				ds_stream_index_iterator_free(sds_it);
+				ds_sds_session_free(session);
+				return OSCAP_ERROR;
+			}
+			_print_single_benchmark_profiles_only(bench);
+
+		} else if (oscap_source_get_scap_type(xccdf_source) == OSCAP_DOCUMENT_XCCDF_TAILORING) {
+			_print_xccdf_tailoring(xccdf_source, 0, _print_xccdf_profile_terse);
+		}
+		ds_sds_session_reset(session);
+	}
+	oscap_string_iterator_free(checklist_it);
+	return OSCAP_OK;
+}
+
+static int app_info_single_ds_one_profile(struct ds_stream_index_iterator* sds_it, struct ds_sds_session *session, const char *profile_id)
+{
+	const char *prefix = "";
+	struct ds_stream_index * stream = ds_stream_index_iterator_next(sds_it);
+	struct oscap_string_iterator* checklist_it = ds_stream_index_get_checklists(stream);
+
+	printf("\nStream: %s\n", ds_stream_index_get_id(stream));
+	printf("Generated: %s\n", ds_stream_index_get_timestamp(stream));
+	printf("Version: %s\n", ds_stream_index_get_version(stream));
+
+	while (oscap_string_iterator_has_more(checklist_it)) {
+		const char * id = oscap_string_iterator_next(checklist_it);
+		// printf("\tRef-Id: %s\n", id);
+
+		/* decompose */
+		struct oscap_source *xccdf_source = ds_sds_session_select_checklist(session, ds_stream_index_get_id(stream), id, NULL);
+		if (xccdf_source == NULL) {
+			oscap_string_iterator_free(checklist_it);
+			ds_stream_index_iterator_free(sds_it);
+			ds_sds_session_free(session);
+			return OSCAP_ERROR;
+		}
+
+		if (oscap_source_get_scap_type(xccdf_source) == OSCAP_DOCUMENT_XCCDF) {
+			struct xccdf_benchmark *bench = xccdf_benchmark_import_source(xccdf_source);
```

Comment:
I think you should free `bench`.

Reply:
```
==9697== 25,510,363 (16 direct, 25,510,347 indirect) bytes in 1 blocks are definitely lost in loss record 657 of 657                                                                                                 
==9697==    at 0x4C2EB6B: malloc (vg_replace_malloc.c:299)                                                
==9697==    by 0x4E8906C: oscap_list_add (list.c:55)                                                      
==9697==    by 0x4EEBA29: xccdf_group_parse (rule.c:228)                                                  
==9697==    by 0x4EEB951: xccdf_content_parse (rule.c:64)                                                 
==9697==    by 0x4EDC40A: xccdf_benchmark_parse (benchmark.c:208)                                         
==9697==    by 0x4EDC56C: xccdf_benchmark_import_source (benchmark.c:69)                                  
==9697==    by 0x40E371: app_info_single_ds_one_profile (oscap-info.c:362)                                
==9697==    by 0x40E371: app_info_single_ds (oscap-info.c:469)                                            
==9697==    by 0x40E371: app_info_sds (oscap-info.c:504)                                                  
==9697==    by 0x40E371: app_info (oscap-info.c:615)                                                      
==9697==    by 0x408857: oscap_module_call (oscap-tool.c:261)                                             
==9697==    by 0x408857: oscap_module_process (oscap-tool.c:346)                                          
==9697==    by 0x40783E: main (oscap.c:80)   
```

Reply:
That's interesting. There was no free before, and if I free the benchmark, I get the segmentation error.
utils/oscap-info.c:261 has a message - `xccdf_benchmark_free not needed, it si already freed by the policy!`, so it is probably better not to free it.

Reply:
you need to use `xccdf_benchmark_free(bench)` instead of just `free(bench)`.

Reply:
Update: There are occasions when it doesn't work - even `xccdf_benchmark_free(bench)` can cause segmentation errors. See `utils/oscap-info.c:614`.

Reply:
@matejak I'm afraid we can't accept code with huge memory leaks or segmentation faults. This must be fixed properly. I think that `valgrind` tool should help us to fix the memory issues. I'll try to look into the problem.